### PR TITLE
Fix #32: add discoveryV5 functions

### LINKED ISSUE #32: add discoveryV5 functions
Adds the following new items to go-waku c-bindings:

## New config items:
- `discV5`: Enable DiscoveryV5.
  Default `false`
- `discV5BootstrapNodes`: Array of bootstrap nodes ENR
   Example (these values can be obtained from fleets.status.im):
```js
   "discV5BootstrapNodes": ["enr:-M-4QCtJKX2WDloRYDT4yjeMGKUCRRcMlsNiZP3cnPO0HZn6IdJ035RPCqsQ5NvTyjqHzKnTM6pc2LoKliV4CeV0WrgBgmlkgnY0gmlwhIbRi9KKbXVsdGlhZGRyc7EALzYobm9kZS0wMS5kby1hbXMzLndha3V2Mi50ZXN0LnN0YXR1c2ltLm5ldAYfQN4DiXNlY3AyNTZrMaEDnr03Tuo77930a7sYLikftxnuG3BbC3gCFhA4632ooDaDdGNwgnZfg3VkcIIjKIV3YWt1Mg8", "enr:-Nm4QLHYoJ5WYQoVzyqPR-pwIeQvi3ONWs-EPwk3uUiBiDseN9Dd7fYbCvkMdeXcuZ-8U9IYdGm38VxSDf_Oq3zZ0cEBgmlkgnY0gmlwhGia74CKbXVsdGlhZGRyc7g6ADg2MW5vZGUtMDEuZ2MtdXMtY2VudHJhbDEtYS53YWt1djIudGVzdC5zdGF0dXNpbS5uZXQGH0DeA4lzZWNwMjU2azGhA1giYsmWV9r2yJZYAiMGHJfjLlLeqAuTAokUGPN__pkxg3RjcIJ2X4N1ZHCCIyiFd2FrdTIP", "enr:-Nm4QC0_ClHzbsutYzgT3jJm7ZY1D4shylAdd6Ac-L4uwAUha1oHM0zwoEkTORVt94W5Cpa0IiyrTcXAYLgpRXpVNUsBgmlkgnY0gmlwhC_y0kmKbXVsdGlhZGRyc7g6ADg2MW5vZGUtMDEuYWMtY24taG9uZ2tvbmctYy53YWt1djIudGVzdC5zdGF0dXNpbS5uZXQGH0DeA4lzZWNwMjU2azGhAhAm-P4q6mWONKcGnbLPU8WXZJ4Qs3AxIbrycvc7PVKsg3RjcIJ2X4N1ZHCCIyiFd2FrdTIP"]
```

- `discV5UDPPort`: UDP port for DiscoveryV5
  Default `9000`

### `extern char* waku_discv5_start()`
Starts the DiscoveryV5 service to discover and connect to new peers

**Returns**

A [`JsonResponse`](#jsonresponse-type).
If the execution is successful, the `result` field is set to `true`.

For example:

```json
{
  "result": true
}
```

### `extern char* waku_discv5_stop()`
Stops the DiscoveryV5 service

**Returns**

A [`JsonResponse`](#jsonresponse-type).
If the execution is successful, the `result` field is set to `true`.

For example:

```json
{
  "result": true
}
```

See: https://github.com/waku-org/go-waku/pull/400

### PR DESCRIPTION
**Pull request by al8n**
Please review this PR.

cc: @richard-ramos @Zeegomo

### CODE REVIEW

File: waku-bindings/src/node/disc_v5.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/// Starts the DiscoveryV5 service to discover and connect to new peers
```
Comment (danielSanchezQ) [inline waku-bindings/src/node/disc_v5.rs:1]: I think the protocol is `discv5` , so we can probably name the module as so.

File: waku-bindings/src/node/config.rs [hunk 2]
```diff
@@ -25,15 +28,31 @@ pub struct WakuNodeConfig {
...
+    #[default(Some(false))]
+    #[serde(rename = "discV5")]
+    pub disc_v5: Option<bool>,
+    /// Array of bootstrap nodes ENR.
+    #[serde(rename = "discV5BootstrapNodes")]
+    pub disc_v5_bootstrap_nodes: Vec<String>,
+    /// UDP port for DiscoveryV5. Default `9000`.
+    #[default(Some(9000))]
+    #[serde(rename = "discV5UDPPort")]
+    pub disc_v5_udp_port: Option<u16>,
```
Comment (danielSanchezQ) [inline waku-bindings/src/node/config.rs:55]: Same here, lets compress to `discv5`

File: waku-bindings/src/node/mod.rs [hunk 3]
```diff
@@ -26,6 +27,7 @@ use crate::general::{
 };
 
 pub use config::{WakuLogLevel, WakuNodeConfig};
+pub use disc_v5::{waku_discv5_start, waku_discv5_stop};
```
Comment (danielSanchezQ) [inline waku-bindings/src/node/mod.rs:30]: This is incomplete. We should encode this within the node itself. Similar to other methods that are only available depending on connection conditions.

**Review by danielSanchezQ — CHANGES_REQUESTED** [review]
Some minor renaming to do.

Those 2 functions should be encoded within the node itself. Ideally would be nice if we can provide an API that do not let confusion about it is available or not. If it turns to be dificcult lets just at least make them part of the node struct itself.

**Comment by danielSanchezQ** [comment]
@al8n 
Could we also add so this methods are used in the example?
Mostly to bump up the code coverage.

**Comment by danielSanchezQ** [comment]
@al8n  It seems there is something going on with the submodule. Did you update it and push the changes pointing to the new main commit containing the discv5 changes in `go-waku`?

**Comment by danielSanchezQ** [comment]
> @al8n 
> Could we also add so this methods are used in the example?
> Mostly to bump up the code coverage.

@al8n is this done? Cannot see it in the changes 🤔

File: waku-bindings/tests/node.rs [hunk 4]
```diff
@@ -29,8 +29,14 @@ pub fn main() -> Result<(), String> {
         min_peers_to_publish: None,
         filter: None,
         log_level: Some(WakuLogLevel::Error),
+        discv5: Some(false),
```
Comment (danielSanchezQ) [inline waku-bindings/tests/node.rs:32]: @al8n , I think it fails because it is not activatedl. Maybe this should be `Some(true)`?

Comment (richard-ramos) [reply waku-bindings/tests/node.rs:32]: Yeah! DiscV5 needs to be enabled for Start() / Stop() to work!

Comment (bacv) [reply waku-bindings/tests/node.rs:32]: Looking at `go-waku` it seems that discoveryV5 is mounted during the `start` of a node. In this case the node is only initialized, but not started, thus `discoveryV5` is not yet mounted, even if the config has it enabled.
It should look something like this:
```
let node = waku_new(Some(config))?;
let node = node.start()?;
println!("Node peer id: {}", node.peer_id()?);

let result = node.discv5_start()?;
println!("Discv5 started: {}", result);
```
But in this case `discv5_start()` method should be implemented for `WakuNodeHandle<Running>` variant. 
I'm not sure if that's how things should look.

**Comment by danielSanchezQ** [comment]
@al8n you can update the submodule commit and use the discv5.stop again:
https://github.com/waku-org/go-waku/pull/402#event-8090083131

**Comment by adv-bacv** [comment]
> @al8n you can update the submodule commit and use the discv5.stop again: [waku-org/go-waku#402 (comment)](https://github.com/waku-org/go-waku/pull/402#event-8090083131)

The referenced link does not resolve this: https://github.com/waku-org/waku-rust-bindings/issues/32#issuecomment-1361776414,
discv5.stop is called internally on node.stop().

**Comment by danielSanchezQ** [comment]
> > @al8n you can update the submodule commit and use the discv5.stop again: [waku-org/go-waku#402 (comment)](https://github.com/waku-org/go-waku/pull/402#event-8090083131)
> 
> The referenced link does not resolve this: [#32 (comment)](https://github.com/waku-org/waku-rust-bindings/issues/32#issuecomment-1361776414), discv5.stop is called internally on node.stop().

Aaaah, didn't saw that. Imo it should be either responsibility of the caller to start/end the service, or completely automated and internally handled when starting/stopping the "node".

**Comment by danielSanchezQ** [comment]
@al8n , the issue of double closing is now [fixed](https://github.com/waku-org/go-waku/pull/413). We just need to keep the settings and remove the api calls as they do not exist anymore (it is handled when starting/stopping the node).
Please do update this PR accordingly :)

**Comment by al8n** [comment]
Hi @richard-ramos, it seems that when enabling the discv5 in settings, the start function will report `Error: "listen udp 0.0.0.0:9000: bind: address already in use"`.

**Comment by richard-ramos** [comment]
> Hi @richard-ramos, it seems that when enabling the discv5 in settings, the start function will report `Error: "listen udp 0.0.0.0:9000: bind: address already in use"`. The details are in the failing CI.

Try setting a random port instead of the default 9000 set in `discV5UDPPort`. Perhaps the CI is already using 9000 for something?

**Comment by al8n** [comment]
> > it seems that when enabling the discv5 in settings, the start function will report `Error: "listen udp 0.0.0.0:9000: bind: address already in use"`. The details are in the failing CI.
> 
> Try setting a random port instead of the default 9000 set in `discV5UDPPort`. Perhaps the CI is already using 9000 for something?

I also try some random port, it reports the same error.

**Comment by danielSanchezQ** [comment]
> > > it seems that when enabling the discv5 in settings, the start function will report `Error: "listen udp 0.0.0.0:9000: bind: address already in use"`. The details are in the failing CI.
> > 
> > 
> > Try setting a random port instead of the default 9000 set in `discV5UDPPort`. Perhaps the CI is already using 9000 for something?
> 
> I also tried some random ports and got the same error.

Doesn't it work on your local machine as well? It does work for me (but I am getting other unrelated errors that we may have to take a look as well).

**Comment by bacv** [comment]
Test fails because `waku-bindings::waku_stop()` method calls `waku_start()` from go-waku internally. Not sure when this was introduced, but I'm testing a fix. There are other errors that are reported if you run a test with `-nocapture`. They do not panic even though it's `expected` to.

**Review by danielSanchezQ — APPROVED** [review]
I like the smell of green CI in the morning 🟢

**Pull request merged**